src/Castle.Core.AsyncInterceptor/InterceptorAdapter.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) 2016 James Skimming. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
+
+namespace Castle.DynamicProxy
+{
+    using System.Reflection;
+    using System.Runtime.CompilerServices;
+    using System.Runtime.ExceptionServices;
+    using System.Threading.Tasks;
+
+    using Castle.DynamicProxy.Invocations;
+
+    /// <inheritdoc cref="IInterceptor"/>
+    internal class InterceptorAdapter : IInterceptor
+    {
+        private static readonly MethodInfo InterceptAsyncFunctionMethodInfo = typeof(InterceptorAdapter).GetMethod(
+            nameof(ExecuteForAsyncFunction),
+            BindingFlags.Instance | BindingFlags.NonPublic);
+
+        private static readonly MethodInfo InterceptFunctionMethodInfo = typeof(InterceptorAdapter).GetMethod(
+            nameof(ExecuteForFunction),
+            BindingFlags.Instance | BindingFlags.NonPublic);
+
+        private readonly IAsyncInterceptor _interceptor;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="InterceptorAdapter"/> class.
+        /// </summary>
+        /// <param name="interceptor">
+        /// The interceptors.
+        /// </param>
+        public InterceptorAdapter(IAsyncInterceptor interceptor)
+        {
+            _interceptor = interceptor;
+        }
+
+        /// <summary>
+        /// Intercepts a method <paramref name="invocation"/>.
+        /// </summary>
+        /// <param name="invocation">The method invocation.</param>
+        public void Intercept(IInvocation invocation)
+        {
+            try
+            {
+                if (invocation.Method.ReturnType == typeof(void))
+                {
+                    _interceptor.InterceptAction(new ActionInvocation(invocation));
+                }
+                else if (invocation.Method.ReturnType == typeof(Task))
+                {
+                    invocation.ReturnValue = _interceptor.InterceptAsyncAction(new AsyncActionInvocation(invocation));
+                }
+                else if (typeof(Task).IsAssignableFrom(invocation.Method.ReturnType))
+                {
+                    invocation.ReturnValue = InterceptAsyncFunctionMethodInfo
+                        .MakeGenericMethod(invocation.Method.ReturnType.GetGenericArguments()[0])
+                        .Invoke(this, new object[] { invocation });
+                }
+                else
+                {
+                    invocation.ReturnValue = InterceptFunctionMethodInfo.MakeGenericMethod(invocation.Method.ReturnType)
```

Comment:
The same issue with reflection

+ some info regarding invoking methods performance penalties
https://codeblog.jonskeet.uk/2008/08/09/making-reflection-fly-and-exploring-delegates/

Reply:
Thank you for the link! Reverted back to `AsyncDeterminationInterceptor`